### PR TITLE
Covariance single row input & null skipping

### PR DESCRIPTION
**Pull request by korowa**
# Which issue does this PR close?

Follow up on #4843 (or #4314)

# Rationale for this change

Covariance function also has an issue with single row inputs processing and cases when inputs contain NULL values

# What changes are included in this PR?

Covariance function now able to process single row and pairs of (NULL, NOT NULL) arguments. As a consequence of proper handling NULLs in Variance, Correlation function now requires additional input filtering - in case one input field is NULL, we should filter out whole record (second field actually) to keep Sttdev (as part of correlation) calculation relevant.

# Are these changes tested?

Fixed/added tests for corresponding accumulators and sqllogictests

# Are there any user-facing changes?

covar_samp/covar_pop/corr functions don't return Errors in case of single row input or (null, not null) argument values

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/covariance.rs [hunk 1]
```diff
@@ -291,19 +291,20 @@ impl Accumulator for CovarianceAccumulator {
...
-            let value2 = arr2.next();
+        for i in 0..values1.len() {
+            let value1 = match values1.is_valid(i) {
+                true => arr1.next(),
+                false => None,
+            };
+            let value2 = match values2.is_valid(i) {
+                true => arr2.next(),
+                false => None,
+            };
```
Comment (ozankabak) [inline datafusion/physical-expr/src/aggregate/covariance.rs:302]: Same as above.
```suggestion
            let value1 = if values1.is_valid(i) {
                arr1.next()
            } else {
                None
            };
            let value2 = if values2.is_valid(i) {
                arr2.next()
            } else {
                None
            };
```

File: datafusion/physical-expr/src/aggregate/covariance.rs [hunk 2]
```diff
@@ -253,33 +253,33 @@ impl Accumulator for CovarianceAccumulator {
...
-
-                self.count += 1;
-                self.mean1 = new_mean1;
-                self.mean2 = new_mean2;
-                self.algo_const = new_c;
+                continue;
             }
+
+            let value1 = unwrap_or_internal_err!(value1);
+            let value2 = unwrap_or_internal_err!(value2);
```
Comment (ozankabak) [inline datafusion/physical-expr/src/aggregate/covariance.rs:271]: Had we migrated to Rust 1.65, I would suggest the much simpler:
```
let Some(value1) = value1 else { continue };
let Some(value2) = value2 else { continue };
```
@alamb, any information as to when we will make the switch?

File: datafusion/physical-expr/src/aggregate/covariance.rs [hunk 3]
```diff
@@ -291,19 +291,20 @@ impl Accumulator for CovarianceAccumulator {
...
+            let value1 = match values1.is_valid(i) {
+                true => arr1.next(),
+                false => None,
+            };
+            let value2 = match values2.is_valid(i) {
+                true => arr2.next(),
+                false => None,
+            };
 
             if value1.is_none() || value2.is_none() {
```
Comment (ozankabak) [inline datafusion/physical-expr/src/aggregate/covariance.rs:304]: Similar to the comment above, if we had Rust 1.65 I would use the `let-else` construct to get rid of the unwraps.

**Review by ozankabak — APPROVED** [review]
I have a few minor comments and suggestions for the future. Other than it looks great. Thank you for the contribution 🚀

File: datafusion/physical-expr/src/aggregate/covariance.rs (hunk 3 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/covariance.rs:304]: Argh - I hate GH's line selection. I am referring to the `continue` check right below.

File: datafusion/physical-expr/src/aggregate/covariance.rs (hunk 2 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/covariance.rs:271]: I am referring to the `is_none`-`continue` check and the `unwrap_or_internal_err!` usage. GH's line selection tool messes up the context -- sorry about that.

File: datafusion/physical-expr/src/aggregate/correlation.rs [hunk 4]
```diff
@@ -145,14 +149,34 @@ impl Accumulator for CorrelationAccumulator {
     }
 
     fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
-        self.covar.update_batch(values)?;
+        let values = if values[0].null_count() != 0 || values[1].null_count() != 0 {
```
[the remark this answers is not in the document]
Comment (korowa) [reply datafusion/physical-expr/src/aggregate/correlation.rs:152]: Unfortunately, I havent found a way to keep this check only in single accumulator

File: datafusion/physical-expr/src/aggregate/covariance.rs (hunk 2 above)
Comment (korowa) [reply datafusion/physical-expr/src/aggregate/covariance.rs]: Looks cleaner, and it seems we dont need this macro anymore

Comment (korowa) [reply datafusion/physical-expr/src/aggregate/covariance.rs]: But yeah, its quite new feature, and we should support 1.62, so I guess, I'll revert this change

Comment (korowa) [reply datafusion/physical-expr/src/aggregate/covariance.rs:273]: Though I still can get rid of unwraps in `retract_batch` using this macro 🤔

File: datafusion/physical-expr/src/aggregate/correlation.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/correlation.rs:152]: Yes, for that we will need to in-line the logic in children accumulator to here. No need to do that in this PR; we can do it in a subsequent PR. However I suggest leaving a `TODO` in the code to remind us.

File: datafusion/physical-expr/src/aggregate/covariance.rs (hunk 2 above)
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/covariance.rs:273]: Agreed, sounds good. We can switch to the new `let-else` construct when the project migrates to 1.65.

File: datafusion/physical-expr/src/aggregate/correlation.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (korowa) [reply datafusion/physical-expr/src/aggregate/correlation.rs:152]: By inlinig you mean not using accumulators themselves, and instead storing all required states in Covariance, updating them with Covariance inputs (just as an example)?

[the remark this answers is not in the document]
Comment (ozankabak) [reply datafusion/physical-expr/src/aggregate/correlation.rs:152]: Yes, that's what I meant

[the remark this answers is not in the document]
Comment (korowa) [reply datafusion/physical-expr/src/aggregate/correlation.rs:152]: Now it seems to me that splitting up input filtering and calculation logic could help to solve these double-checks

[the remark this answers is not in the document]
Comment (korowa) [reply datafusion/physical-expr/src/aggregate/correlation.rs:157]: Left a note for future us

**Comment by alamb** [comment]
> Had we migrated to Rust 1.65, I would suggest the much simpler:
> @alamb, any information as to when we will make the switch?

I am not sure what you are asking here @ozankabak . The CI checks use the latest rust stable version. I just verified that this is currently using Rust `1.66`. 

https://github.com/apache/arrow-datafusion/actions/runs/3878200568/jobs/6614116891#step:4:97

**Comment by alamb** [comment]
Thanks @korowa  and @ozankabak

**Comment by ozankabak** [comment]
@alamb I see `rust-version = "1.62"` in lots of TOML files, so I presumed that's the Rust version we are committing to support at the moment.

**Comment by ozankabak** [comment]
OK, I will assume we are using the latest Rust. There is some good syntax that makes cumbersome things more readable there -- would definitely like to leverage them.

**Comment by alamb** [comment]
Just to be clear @ozankabak  -- I think we use the latest **stable** rust (not nightly).

**Pull request merged**